### PR TITLE
Fix #19245: NPE in UpdateRideApproachExitWaypoints()

### LINKED ISSUE #19245: [905974] openrct2.exe: Guest::UpdateRideApproachExitWaypoints
<table>
    <tr>
        <td></td>
        <td><b><a href="https://openrct2.sp.backtrace.io/dashboard/openrct2/project/openrct2/group/905974fbb6f4797a49a504844dcc1274bc3f3c11158f6def330dae9f3ace313d">New error: [openrct2]</a></b><br /><b>1</b> occurrence since (Fri Nov 15 2019 11:27:39)<br /></td>
    </tr>
</table><table>
    <tr><td><strong>Application</strong></td><td>openrct2.exe</td></tr>
<tr>
    <td><strong>Attributes</strong></td>
    <td>
<strong>_rxid</strong> &mdash; 1
        <ul>
                <li><code>7e000000-484c-7e09-0000-000000000000</code> &mdash; 1</li></ul><br>
<strong>attachment_config.ini</strong> &mdash; 1
        <ul>
                <li><code>C:\Users\James\Documents\OpenRCT2\crash\df7473f8-0e72-407b-a39d-0bc26dacff65.ini</code> &mdash; 1</li></ul><br>
<strong>assert_failure</strong> &mdash; 1
        <ul>
                <li><code>*</code> &mdash; 1</li></ul><br>
<strong>fault.address</strong> &mdash; 1
        <ul>
                <li><code>344</code> &mdash; 1</li></ul><br>
<strong>version</strong> &mdash; 1
        <ul>
                <li><code>OpenRCT2, v0.4.3-183-g09b1074 (09b1074 on develop) provided by GitHub</code> &mdash; 1</li></ul><br>
<strong>error.message</strong> &mdash; 1
        <ul>
                <li><code>EXCEPTION_ACCESS_VIOLATION_READ</code> &mdash; 1</li></ul><br>
<strong>commit</strong> &mdash; 1
        <ul>
                <li><code>09b1074</code> &mdash; 1</li></ul><br></td>
    </tr>
    <tr>
        <td><strong>Classifiers</strong></td>
        <td>
            <strong>invalid-read</strong><br></td>
    </tr>
    <tr>
        <td><strong>Frames</strong></td>
        <td>
            <code>
Guest::UpdateRideApproachExitWaypoints</code>
        </td>
    </tr></table>

### PR DESCRIPTION
**Pull request by Gymnasiast**

### CODE REVIEW

File: src/openrct2/entity/Guest.cpp [hunk 1]
```diff
@@ -4513,16 +4513,26 @@ void Guest::UpdateRideApproachExitWaypoints()
...
-        Guard::Assert((Var37 & 3) < 3);
-        targetLoc.x += carEntry->peep_loading_waypoints[Var37 / 4][Var37 & 3].x;
-        targetLoc.y += carEntry->peep_loading_waypoints[Var37 / 4][Var37 & 3].y;
+        const CarEntry& carEntry = rideEntry->Cars[vehicle->vehicle_type];
 
+        const size_t fours = Var37 / 4;
+        if (fours >= carEntry.peep_loading_waypoints.size())
+            return;
+
+        const auto remainder = Var37 & 3;
```
Comment (duncanspumpkin) [inline src/openrct2/entity/Guest.cpp:4529]: This is normally called the waypoint in the rest of the code.

File: src/openrct2/entity/Guest.cpp [hunk 2]
```diff
@@ -4513,16 +4513,26 @@ void Guest::UpdateRideApproachExitWaypoints()
...
+
+        if (vehicle->vehicle_type >= std::size(rideEntry->Cars))
+            return;
 
-        Guard::Assert((Var37 & 3) < 3);
-        targetLoc.x += carEntry->peep_loading_waypoints[Var37 / 4][Var37 & 3].x;
-        targetLoc.y += carEntry->peep_loading_waypoints[Var37 / 4][Var37 & 3].y;
+        const CarEntry& carEntry = rideEntry->Cars[vehicle->vehicle_type];
 
+        const size_t fours = Var37 / 4;
```
Comment (duncanspumpkin) [inline src/openrct2/entity/Guest.cpp:4525]: I think we could name this `carPosition` or `carNumber`

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
Update the var names and this is good to go.

File: src/openrct2/entity/Guest.cpp (hunk 2 above)
Comment (tupaschoal) [reply src/openrct2/entity/Guest.cpp:4525]:
```suggestion
        const size_t carPosition = Var37 / 4;
```

File: src/openrct2/entity/Guest.cpp [hunk 3]
```diff
@@ -4513,16 +4513,26 @@ void Guest::UpdateRideApproachExitWaypoints()
...
+        if (fours >= carEntry.peep_loading_waypoints.size())
```
Comment (tupaschoal) [inline src/openrct2/entity/Guest.cpp:4526]:
```suggestion
        if (carPosition >= carEntry.peep_loading_waypoints.size())
```

File: src/openrct2/entity/Guest.cpp (hunk 1 above)
Comment (tupaschoal) [reply src/openrct2/entity/Guest.cpp:4529]:
```suggestion
        const auto waypoint = Var37 & 3;
```

File: src/openrct2/entity/Guest.cpp [hunk 4]
```diff
@@ -4513,16 +4513,26 @@ void Guest::UpdateRideApproachExitWaypoints()
...
+            return;
+
+        const auto remainder = Var37 & 3;
+        Guard::Assert(remainder < 3);
```
Comment (tupaschoal) [inline src/openrct2/entity/Guest.cpp:4530]:
```suggestion
        Guard::Assert(waypoint < 3);
```

File: src/openrct2/entity/Guest.cpp [hunk 5]
```diff
@@ -4513,16 +4513,26 @@ void Guest::UpdateRideApproachExitWaypoints()
...
+
+        const auto& rtd = ride->GetRideTypeDescriptor();
+
+        CoordsXY targetLoc = rtd.GetGuestWaypointLocation(*vehicle, *ride, CurrentRideStation);
+        targetLoc += carEntry.peep_loading_waypoints[fours][remainder];
```
Comment (tupaschoal) [inline src/openrct2/entity/Guest.cpp:4535]:
```suggestion
        targetLoc += carEntry.peep_loading_waypoints[carPosition][waypoint];
```

**Review by tupaschoal — COMMENTED** [review]
@duncanspumpkin see if the changes I suggested are what you meant, if so I'll add them in batch to merge.

**Comment by duncanspumpkin** [comment]
> @duncanspumpkin see if the changes I suggested are what you meant, if so I'll add them in batch to merge.

Yeah that looks good

**Review by duncanspumpkin — APPROVED** [review]

**Comment by tupaschoal** [comment]
The checks here have failed to report, somehow. They gave all finished, but still show up as pending, yet they can't be cancelled since you can't cancel a complete job. I'll have to do some dummy diff and then revert to retrigger them.

File: src/openrct2/entity/Guest.cpp [hunk 6]
```diff
@@ -4513,16 +4513,26 @@ void Guest::UpdateRideApproachExitWaypoints()
             return;
         }
 
-        const auto& rtd = ride->GetRideTypeDescriptor();
-        CoordsXY targetLoc = rtd.GetGuestWaypointLocation(*vehicle, *ride, CurrentRideStation);
-
         RideObjectEntry* rideEntry = vehicle->GetRideEntry();
-        CarEntry* carEntry = &rideEntry->Cars[vehicle->vehicle_type];
+        if (rideEntry == nullptr)
```
Comment (tupaschoal) [inline src/openrct2/entity/Guest.cpp:4517]:
```suggestion
        if (rideEntry == nullptr )
```

File: src/openrct2/entity/Guest.cpp [hunk 7]
```diff
@@ -4513,16 +4513,26 @@ void Guest::UpdateRideApproachExitWaypoints()
             return;
         }
 
-        const auto& rtd = ride->GetRideTypeDescriptor();
-        CoordsXY targetLoc = rtd.GetGuestWaypointLocation(*vehicle, *ride, CurrentRideStation);
-
         RideObjectEntry* rideEntry = vehicle->GetRideEntry();
-        CarEntry* carEntry = &rideEntry->Cars[vehicle->vehicle_type];
+        if (rideEntry == nullptr )
```
Comment (tupaschoal) [inline src/openrct2/entity/Guest.cpp:4517]:
```suggestion
        if (rideEntry == nullptr)
```

**Pull request merged**